### PR TITLE
Add document which describes how to build each of the calico components

### PR DESCRIPTION
**Pull request by robbrockbank**
Added doc to describe how to build the various Calico components - primary focus on k8s deployment (KDD)

### CODE REVIEW

File: BUILDING_CALICO.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+To build Calico CNI plugins, clone the felix repo and checkout the correct tag.
+
+```
+cd $BASEDIR
+git clone git@github.com:projectcalico/cni-plugin.git
+cd cni-plugin
+git checkout tags/$VERSION_CNI
+```
+
+Build the `calico/felix` docker image (which also builds the associated binaries) and
```
Comment (caseydavenport) [inline BUILDING_CALICO.md]: I think you've got a copy/paste error here - this isn't Felix we're building :)

File: BUILDING_CALICO.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+make docker-image
+docker tag calico/felix calico/felix:$VERSION_FELIX
+```
+
+#### What has this built?
+
+-  The `calico/cni` Docker image
+-  The `dist/calico` binary (included in the Docker image)
+-  The `dist/calico-ipam` binary (included in the Docker image)
+-  The `dist/flannel` binary (included in the Docker image)
```
Comment (caseydavenport) [inline BUILDING_CALICO.md]: Strictly speaking, the things below this are not built here - they are pulled in from external sources pre-built.

File: BUILDING_CALICO.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+#### What has this built?
+
+-  The `dist/bird` binary
+-  The `dist/bird6` binary
+-  The `dist/birdcl` binary
+
+### 7. libnetwork-plugin (used in the calico/node image)
+
+To build the libnetwork-plugin binaries, clone the bird repo and checkout the correct tag.
```
Comment (caseydavenport) [inline BUILDING_CALICO.md]: "clone the bird repo"

File: BUILDING_CALICO.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+Build the `calico/libnetwork-plugin` container.
+
+```
+make clean
+make calico/libnetwork-plugin
+```
+
+#### What has this built?
+
+-  The `calico/libnetwork-plugin` binary
```
Comment (caseydavenport) [inline BUILDING_CALICO.md]: binary -> docker image

File: BUILDING_CALICO.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+git checkout tags/$VERSION_CNI
+```
+
+Build the `calico/cni` docker image (which also builds the associated binaries) and
+retag the image with the correct version.
+
+```
+make clean
+make docker-image
+docker tag calico/felix calico/felix:$VERSION_FELIX
```
Comment (caseydavenport) [inline BUILDING_CALICO.md]: This should be `calico/cni`

File: BUILDING_CALICO.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+Build the `calico/typha` docker image (which also builds the associated binaries) and
+retag the image with the correct version.
+
+```
+make clean
+make calico/typha
+docker tag calico/typha calico/typha:$VERSION_TYPHA
+```
+
+#### What has this built?
```
Comment (ozdanborne) [inline BUILDING_CALICO.md]: Could we just call this `Artifacts`?

**Comment by ozdanborne** [comment]
Can we add links between this document and repo-structure?
http://docs.projectcalico.org/v2.2/reference/repo-structure

**Comment by ozdanborne** [comment]
If this is going to be a raw markdown doc please add it to excludes:
https://github.com/projectcalico/calico/blob/master/_config.yml#L10

File: BUILDING_CALICO.md (hunk 5 above)
Comment (robbrockbank) [inline BUILDING_CALICO.md]: grr  ... ta

File: BUILDING_CALICO.md (hunk 6 above)
Comment (robbrockbank) [inline BUILDING_CALICO.md]: Yeah we can do, and since some of the artifacts have not been built it might be a better heading.

**Comment by caseydavenport** [comment]
> http://docs.projectcalico.org/v2.2/reference/repo-structure

Personally I think there's a good case for removing that doc from the docs.projectcalico.org site and combining it with this one, or turning it into its own markdown file.

I think developer-focused docs shouldn't go in docs.projectcalico.org, just user-focused docs.

**Comment by robbrockbank** [comment]
@ozdanborne :  I tend to agree that I'm not sure I like the Repo Structure doc being in the docs (or at least not in it's current guise as it's not really anything to do with repo "structure").

**Comment by robbrockbank** [comment]
@ozdanborne @caseydavenport  final review?

**Pull request merged**